src/time_driver.rs

```diff
@@ -0,0 +1,246 @@
+}
+
+embassy_time_driver::time_driver_impl!(static DRIVER: MultiFunctionTimerDriver = MultiFunctionTimerDriver {
+    period: AtomicU32::new(0),
+    alarm: Mutex::const_new(CriticalSectionRawMutex::new(), AlarmState::new()),
+    queue: Mutex::new(RefCell::new(Queue::new()))
+});
+
+impl MultiFunctionTimerDriver {
+    fn init(&'static self, _cs: critical_section::CriticalSection) {
+        let r = regs();
+
+        if TICK_HZ != 32768 {
+            panic!("Only a tickrate of 32KHz is supported");
+        }
+
+        unsafe { enable_interrupt() };
+
+        // Disable the clocks.
+        r.tn_ckc()
+            .write(|w| unsafe { w.c1csel().bits(0b000).c2csel().bits(0b000) });
+
+        // Configure Mode 3: Dual-Independent Timers.
+        r.tn_mctrl().write(|w| unsafe { w.mdsel().bits(0b010) });
+
+        // Reset the counters to the same value.
+        r.tn_cnt1().write(|w| unsafe { w.bits(0xffff) });
+        r.tn_cnt2().write(|w| unsafe { w.bits(0xffff) });
+
+        // Set both counters to be overflowed (reload-captured) to the same value.
+        r.tn_cra().write(|w| unsafe { w.bits(0xffff) });
+        r.tn_crb().write(|w| unsafe { w.bits(0xffff) });
+
+        // Set Counter 1 to compare-match at the half-way point.
+        // Note: we'll use Counter 2 as alarm and thus will not need to configure it now.
+        r.tn_cpa().write(|w| unsafe { w.bits(0x7fff) });
+
+        // Set Counter comparison to equality.
+        r.tn_cpcfg().write(|w| w.eqaen().set_bit().eqben().set_bit());
+
+        // Enable overflow and half overflow detection interrupts.
+        r.tn_ien().write(|w| w.taien().set_bit().teien().set_bit());
+
+        // Set low power clock mode before starting the clocks.
+        r.tn_ckc().write(|w| w.low_pwr().bit(true));
+
+        // Starts the clocks sources from LFCLK.
+        r.tn_ckc()
+            .modify(|_, w| unsafe { w.c1csel().bits(0b100).c2csel().bits(0b100) });
+    }
+
+    fn on_interrupt(&self) {
+        let r = regs();
+        let pending = r.tn_ectrl().read();
+
+        critical_section::with(|cs| {
+            // Clear pending interrupts.
+            r.tn_eclr().write(|w| unsafe { w.bits(pending.bits()) });
+
+            // Counter 1 - reloaded (overflow)
+            if pending.tapnd().bit_is_set() {
+                self.next_period(cs);
+            }
+
+            // Counter 1 - compare matched (half overflow)
+            if pending.tepnd().bit_is_set() {
+                self.next_period(cs);
+            }
+
+            // Counter 2 - compare matched (alarm)
+            if pending.tfpnd().bit_is_set() {
+                self.trigger_alarm(cs);
+            }
+        });
+    }
+
+    fn next_period(&self, cs: CriticalSection) {
+        // We only modify the period from the timer interrupt, so we know this can't race.
+        let period = self.period.load(Ordering::Relaxed) + 1;
+        self.period.store(period, Ordering::Relaxed);
```

Comment:
Hmmm... Did you copy this from somewhere? It's in a critical section, so it's fine. But our chip supports full atomics. This could be replaced by one `fetch_add`

Reply:
Yes, I took heavy inspiration from the `embassy-stm32` `time_driver` implementation.

In this case it is either what I committed, or
```rust
let period = self.period.fetch_add(1, Ordering::Relaxed) + 1;
```
Which in some sense is more operations. I don't mind either way.